editor/modes/visual-editor/block-list.js

```diff
@@ -23,14 +24,56 @@ class VisualEditorBlockList extends wp.element.Component {
 		this.onSelectionStart = this.onSelectionStart.bind( this );
 		this.onSelectionChange = this.onSelectionChange.bind( this );
 		this.onSelectionEnd = this.onSelectionEnd.bind( this );
+		this.onPointerMove = throttle( this.onPointerMove.bind( this ), 250 );
+		// Browser does not fire `*move` event when the pointer position changes
+		// relative to the document, so fire it with the last known position.
+		this.onScroll = () => this.onPointerMove( { clientY: this.lastClientY } );
+		this.setBlockRef = this.setBlockRef.bind( this );
 
 		this.state = {
 			selectionAtStart: null,
 		};
+
+		this.coordMap = {};
+		this.coordMapKeys = [];
+		this.lastClientY = 0;
+		this.refs = {};
+	}
+
+	componentDidUpdate() {
+		// Clean up refs.
+		this.refs = pickBy( this.refs );
+	}
+
+	setBlockRef( ref, uid ) {
+		this.refs = {
+			...this.refs,
+			[ uid ]: ref,
+		};
+	}
+
+	onPointerMove( { clientY } ) {
+		const y = clientY + window.pageYOffset;
+		const key = this.coordMapKeys.reduce( ( acc, topY ) => y > topY ? topY : acc );
+
+		this.lastClientY = clientY;
+		this.onSelectionChange( this.coordMap[ key ] );
 	}
 
 	onSelectionStart( uid ) {
+		const pageYOffset = window.pageYOffset;
```

Comment:
Minor (and sometimes I question its readability): This could be simplified with destructuring to:

```js
const { pageYOffset } = window;
```